### PR TITLE
Allow specifying which backend we want to use for tests and in config

### PR DESCRIPTION
**Pull request by jder**
Code previously assumed if you had a GPU you wanted to have distributed = true and (AFAICT) have the right env vars set for that. This tries to let you pick if you want cpu, gpu, or distributed. (Or try to detect automatically.) 

This now lets us run tests on CPUs in CI. Next we'll set up GPU tests and run both.

### CODE REVIEW

File: configs/train_cm4.test.yaml [hunk 1]
```diff
@@ -25,14 +25,7 @@ inference:
...
 step_transition: []
-
-distributed:
-  enabled: true
-  dist_url: null
-  world_size: null
-  rank: null
-  gpu: null
-  dist_backend: null
+backend: cpu
```
Comment (jder) [inline configs/train_cm4.test.yaml]: Having this be `cpu` rather than `auto` is basically the fix to run tests (for now) on cpu. Previously the `distributed` blocks were unused (they were populated in code) except for the 'enabled' boolean. You can now control that with this backend option. (`nccl` vs `cuda` or `cpu`)

File: src/aggregator/inference/main.py [hunk 2]
```diff
@@ -19,6 +19,7 @@ class InferenceEvaluatorAggregator:
 
     def __init__(
         self,
+        device: torch.device,
```
Comment (jder) [inline src/aggregator/inference/main.py]: Most of the changes in this PR are passing the device around. Not sure how much this is worth it vs setting the global device using `with torch.device(…)` -- do we want tensors to be allocated on the CPU by default and only moved to GPU manually with `to`?

File: src/backend.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        case "auto" if torch.cuda.is_available():
+            logging.info("auto backend detected CUDA")
+            device = torch.device("cuda")
+        case "auto":
+            logging.info("auto backend: cuda not found, using CPU")
+            device = torch.device("cpu")
+        case _:
+            raise ValueError(f"Invalid backend: {backend}")
+
+    return device
```
Comment (jder) [inline src/backend.py:66]: This file now contains the logic to decide which device & distribution config to use based on the `backend` config option. These are run once and then the device is passed around rather than a global `get_device`.

File: src/datasets.py [hunk 4]
```diff
@@ -76,9 +74,8 @@ def __init__(
         self.wet_surface = wet_surface.bool()
         self.size = len(self.rolling_indices)
 
-        if using_gpu():
-            self.wet = self.wet.pin_memory()
-            self.wet_surface = self.wet_surface.pin_memory()
+        self.wet = self.wet.pin_memory()
+        self.wet_surface = self.wet_surface.pin_memory()
```
Comment (jder) [inline src/datasets.py]: I think this is harmless?

File: src/stepper.py [hunk 5]
```diff
@@ -26,12 +26,11 @@ def train_step(
     @staticmethod
     @torch.no_grad()
     def validate_step(
-        model: torch.nn.Module, batch: TrainData, loss_fn: Callable
+        model: BaseModel, batch: TrainData, loss_fn: Callable
     ) -> ValOutput:
         assert len(batch) == 1  # Assert we are using one step of input and output
         input = batch.get_input(0)
         label = batch.get_label(0)
-        model = model.module if using_gpu() else model
```
Comment (jder) [inline src/stepper.py:34]: This is actually a bug at the moment. The issue is that when using GPU we wrap the BaseModel into a `DistributedDataParallel` module. But that doesn't expose `forward_once`, so we unwrap here. We can of course pass the device along here and use the same conditional logic but I wonder if we could remove that by moving the logic currently in BaseModel.forward into Stepper and having UNet.forward be what's currently in forward_once. Is that better? Worse?

File: src/train_3D.py [hunk 6]
```diff
@@ -541,7 +536,7 @@ def inference_one_epoch(self, epoch):
             )
 
             Stepper.inference(
-                model=self.model.module if using_gpu() else self.model,
+                model=self.model,
```
Comment (jder) [inline src/train_3D.py]: This is also a bug for the same reason.

File: src/train_3D.py [hunk 7]
```diff
@@ -736,14 +731,14 @@ def load_checkpoint(self, checkpoint_path, finetune=False):
             logging.info(f"Optimizer LR: {self.optimizer.param_groups[-1]['lr']}")
 
     def is_wandb_enabled(self):
-        if using_gpu():
-            return self.wandb_logger.enabled and is_main_process()
-        else:
-            return self.wandb_logger.enabled
+        return self.wandb_logger.enabled and is_main_process()
```
Comment (jder) [inline src/train_3D.py:736]: is_main_process() already correctly handles the non-gpu case, I think?

File: src/datasets.py (hunk 4 above)
Comment (jder) [reply src/datasets.py]: Ok, CI tells me it's not harmless, that's too bad.

File: src/aggregator/inference/main.py (hunk 2 above)
Comment (alxmrs) [reply src/aggregator/inference/main.py]: > do we want tensors to be allocated on the CPU by default and only moved to GPU manually with to?

A great question! I think we'll have to decide based on a performance investigation. My intuition says "yes", though.

File: src/backend.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import torch
+
+from config import DistributedConfig, EvalBackendConfig, TrainBackendConfig
+from utils.distributed import init_distributed_mode
+
+
+def init_train_backend(
+    backend: TrainBackendConfig,
+) -> tuple[torch.device, DistributedConfig | None]:
+    match backend:
```
Comment (alxmrs) [inline src/backend.py:14]: Nice use of pattern matching!

File: src/backend.py (hunk 3 above)
Comment (alxmrs) [reply src/backend.py:66]: This explanation is great! I think it would be nice to have as a docstring at the top of the file.

File: src/datasets.py (hunk 4 above)
Comment (alxmrs) [reply src/datasets.py]: lool!

File: src/stepper.py (hunk 5 above)
Comment (alxmrs) [reply src/stepper.py:34]: Great questions, I haven't looked at that part of the codebase yet, but hope to soon.

File: tests/conftest.py [hunk 9]
```diff
@@ -40,9 +39,13 @@ def model2_path(request):
 
 
 # Run a test for both CPU and GPU, and allows selecting or skipping CUDA tests.
-@pytest.fixture(params=["cpu", pytest.param("cuda", marks=pytest.mark.cuda)])
-def device(request):
-    return torch.device(request.param)
+# TODO(jder): At the moment, we can't use this because Trainer can only be
+# inited once per process. Once that is we should accept it in `train_config` below.
+@pytest.fixture(
```
Comment (alxmrs) [inline tests/conftest.py:52]: Thanks for this note. I like the change that you've made here.

I'm not sure if this is useful, but we could mark the scope for this fixture to be `session` and then it should only call this function once per testing session (per parameter). That's how I've gotten the trainer instantiation to work thus far. 

I think this comment already alludes to this, but: I like the idea of letting this fixture set the device in the train_config.

File: src/stepper.py (hunk 5 above)
Comment (suryadheeshjith) [reply src/stepper.py:34]: > But that doesn't expose forward_once, so we unwrap here. We can of course pass the device along here and use the same conditional logic but I wonder if we could remove that by moving the logic currently in BaseModel.forward into Stepper and having UNet.forward be what's currently in forward_once.

This is a great catch. I am planning on doing exactly the same (#51) but what stopped me from doing this before were some issues with how batch normalization is implemented in pytorch alongside torch's DistributedDataParallel module. See more [here](https://github.com/pytorch/pytorch/issues/66504#issuecomment-1007806934). I also created a notebook that describes this issue [here](https://github.com/suryadheeshjith/Ocean_Emulator/blob/main/notebooks/batchnorm_test.ipynb).

File: src/backend.py (hunk 3 above)
Comment (jder) [reply src/backend.py:66]: Added it to the relevant functions.

File: src/stepper.py (hunk 5 above)
Comment (jder) [reply src/stepper.py:34]: ok, for now I replaced this with a type test instead of passing the device around which seems more direct anyway.

File: tests/conftest.py (hunk 9 above)
Comment (jder) [reply tests/conftest.py:52]: Yeah, makes sense. Right now it is already session scoped but since it yields two configurations it runs into the same issue.

File: src/aggregator/inference/main.py (hunk 2 above)
Comment (jder) [reply src/aggregator/inference/main.py]: OK, I left behavior unchanged for now.

**Comment by jder** [comment]
ok, I think this is ready for a review. My main outstanding question: I think this probably breaks one or more notebooks since you must now pass a device to some classes you didn't have to before. @suryadheeshjith which notebooks should we be 
maintaining/validating as we make changes like this?

**Comment by jder** [comment]
This is ready for another look. As discussed this no longer hands the device around -- I left a TODO with a link to the issue. We can now run tests on either CPU or GPU but not both together. 

@suryadheeshjith please let me know about this too

> @suryadheeshjith which notebooks should we be maintaining/validating as we make changes like this?

**Comment by suryadheeshjith** [comment]
You may ignore notebooks for now. For future reference, the main notebook I use for evaluation is Samudra_Eval.ipynb. You can ignore it for now.

**Review by suryadheeshjith — APPROVED** [review]
LGTM

File: src/train_3D.py [hunk 10]
```diff
@@ -31,21 +31,16 @@
 from constants import EXTRA_VARS, INPT_VARS, OUT_VARS, TensorMap, construct_metadata
 from datasets import InferenceDataset, InferenceDatasets, TrainDataset
 from models.unet import UNet
+from src.backend import init_train_backend
```
Comment (alxmrs) [inline src/train_3D.py]: 🛑  I'm not a fan of including `src.` in the import path.

File: src/utils/device.py [hunk 11]
```diff
@@ -1,11 +1,28 @@
...
+
 
 def using_gpu() -> bool:
     return get_device().type == "cuda"
 
 
+# TODO(jder): We'd like to remove this (and other global singletons)
+# by moving to some training framework + context objects once we have
+# benchmarking set up.
+def set_device(device: torch.device) -> None:
```
Comment (alxmrs) [inline src/utils/device.py:13]: These LGTM, thanks!

**Review by alxmrs — APPROVED** [review]
I have one (small) blocker, but otherwise, this PR LGTM. After it's fixed, this looks ready to merge.

File: src/train_3D.py (hunk 10 above)
Comment (alxmrs) [reply src/train_3D.py]: Is there some way that you can make the import like `constants` and `datasets` above?

Comment (jder) [reply src/train_3D.py]: This (and a few others I found) were added automatically by ruff. I think this is because `src` has an `__init__.py` in it and ideally we'd make it `src/samudra/__init__.py` (and maybe have tests live inside samudra as Surya said elsewhere). I didn't see a way to configure ruff otherwise (I tried a couple things like setting src = "src").

Comment (alxmrs) [reply src/train_3D.py]: That makes sense! Thanks for letting me know. There may be a interesting `pyproject.toml` config we could add to make this work, but I agree that `src/samudra/` packaging is the way to go. For a future PR, ofc.
